[nicrandomlee] iP

nicrandomlee opened the pull request:
# CupidBot

> Yesterday you said tomorrow. 
> Just DO IT! ([source](https://en.wikipedia.org/wiki/Cristiano_Ronaldo))


CupidBot is your one stop task tracker! It's
- Text-based
- Fast
- _Easy to use_

All you need to do is the following
1. Download from this [link](https://github.com/nus-cs2103-AY2324S1/ip/pull/97)
2. Double-click it to run
3. Add your tasks.
4. Get reminded, all day and every day 

And it is **FREE!** 😄 

These are the things you can do with it:
- [ ] Add ToDos
- [ ] Add Deadlines
- [ ] Add Events
And **more!**

Example of Cupid Commands are:
```
todo water flowers
```
```
todo water flowers added!
```

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

nicrandomlee committed:
(a) Give your chatbot a new name, to differentiate it from the placeholder name Duke.

(b) Implement an initial skeletal version of the it that simply greets the user and exits.
Example:

nicrandomlee committed:
Improve the skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye.

nicrandomlee committed:
Level 2. Add, List
Add the ability to store whatever text entered by the user and display them back to the user when requested.

nicrandomlee committed:
Level 3. Mark as Done
Add the ability to mark tasks as done. Added the ability to change the status back to not done.

nicrandomlee committed:
ToDos: tasks without any date/time attached to it e.g., visit new theme park
Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
Events: tasks that start at a specific date/time and ends at a specific date/time
e.g., (a) team project meeting 2/10/2019 2-4pm (b) orientation week 4/10/2019 to 11/10/2019

nicrandomlee committed:
Level 5. Handle Errors
Teach the chatbot to deal with errors such as incorrect inputs entered by the user.

nicrandomlee committed:
Level 6. Delete
Add support for deleting tasks from the list.

nicrandomlee committed:
Patched main function

nicrandomlee committed:
Added more errors handled

nicrandomlee committed:
Modified delete function, added more comprehensive automated test case

nicrandomlee committed:
-

nicrandomlee committed:
Added save feature

nicrandomlee committed:
Implemented Load/Save feature upon booting/shutdown

nicrandomlee committed:
add dates and times

nicrandomlee committed:
Add Storage and TaskList

nicrandomlee committed:
Edit Storage

nicrandomlee committed:
Add UI

nicrandomlee committed:
update parser

nicrandomlee committed:
Add LocalDateTime

nicrandomlee committed:
Add Find

nicrandomlee committed:
edit save

nicrandomlee committed:
javadoc for save

nicrandomlee committed:
added save javadoc

nicrandomlee committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/Save.java

nicrandomlee committed:
packaged files

nicrandomlee committed:
reordered cupid.java

nicrandomlee committed:
Added Javadoc

Blizzeracz commented on src/main/java/functions/Load.java:
```diff
@@ -0,0 +1,102 @@
+                    break;
+
+                case "E":
+                    descriptionEndIndex = currentTaskAsString.indexOf("(from:")-1;
```
Maybe you could create a new class or transfer these codes to their respective classes? 🤔 I realized the same issue in other methods in parse too!

Avoid long methods as they often contain more information than what the reader can process at a time. Take corrective action when it goes beyond 30 LoC. The bigger the haystack, the harder it is to find a needle.

Blizzeracz commented on src/main/java/functions/Parser.java:
```diff
@@ -0,0 +1,223 @@
+        this.taskList = taskList;
+    }
+
+    public void parse() {
```
Very neat!

Blizzeracz commented on src/main/java/functions/Parser.java:
```diff
@@ -0,0 +1,223 @@
+
+    public void deleteCommandHandler(String[] inputArray, TaskList taskList) {
+        try {
+            int targetTaskIdx = Integer.parseInt(inputArray[1]) -1;
+            Task task = taskList.get(targetTaskIdx);
+            taskList.remove(targetTaskIdx);
+            System.out.println("Noted: I've removed this task:");
+            System.out.println(task.getTaskAsString());
```
Maybe each logical units within a block should be separated by one blank line?

```suggestion
            // Get the task and task index
            int targetTaskIdx = Integer.parseInt(inputArray[1]) -1;
            Task task = taskList.get(targetTaskIdx);
            
            taskList.remove(targetTaskIdx);
            
            System.out.println("Noted: I've removed this task:");
            System.out.println(task.getTaskAsString());
```

Rationale: Enhances readability by introducing white space between logical units. Each block is often introduced by a comment.

Blizzeracz commented on src/main/java/functions/Parser.java:
```diff
@@ -0,0 +1,223 @@
+            int firstSpaceIndex = input.indexOf(" ");
+            String secondHalfInput = input.substring(firstSpaceIndex+1);
+
+            switch (function) {
+                case LIST:
+                    listCommandHandler(taskList);
+                    break;
+
+                case MARK:
+                    markCommandHandler(inputArray, taskList);
+                    break;
+
+                case UNMARK:
+                    unmarkCommandHandler(inputArray, taskList);
+                    break;
+
+                case DELETE:
+                    deleteCommandHandler(inputArray, taskList);
+                    break;
+
+                case TODO:
+                    toDoCommandHandler(taskList, secondHalfInput);
+                    break;
+
+                case DEADLINE:
+                    deadlineCommandHandler(taskList, secondHalfInput);
+                    break;
+
+                case EVENT:
+                    eventCommandHandler(taskList, input, secondHalfInput);
+                    break;
+
+                case FIND:
+                    findCommandHandler(taskList, secondHalfInput);
+
+                default:
+                    break;
+            }
```
```suggestion
switch (condition) {
case ABC:
    statements;
    // Fallthrough
case DEF:
    statements;
    break;
case XYZ:
    statements;
    break;
default:
    statements;
    break;
}
```

Note there is no indentation for case clauses.

Blizzeracz reviewed (commented):
Overall, I found your code easy to read for the most part except a few places
where the methods were too long, I think those can be abstracted out. I noted a few minor coding standard violations too. The ordering of import statement are all consistent though and the rest of the classes are quite neat. 👍

Blizzeracz commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,45 @@
+
+public class Deadline extends Task {
+
+    private String TaskIcon = "[D]";
```
```suggestion
    private String taskIcon = "[D]";
```

 Variable names must be in camelCase.

Blizzeracz commented on src/main/java/functions/Save.java:
```diff
@@ -0,0 +1,64 @@
+import java.nio.file.StandardCopyOption;
+
+public class Save {
+    TaskList taskList;
```
```suggestion
    private TaskList taskList;
```

Missing modifier

Blizzeracz reviewed (commented):
Adding on to the previous comment, just  small mistakes.

xavierpok commented on src/main/java/functions/Load.java:
```diff
@@ -0,0 +1,102 @@
+        while (s.hasNext()) {
+            String currentTaskAsString = s.nextLine();
+            String taskType = currentTaskAsString.substring(1,2).toUpperCase();
+            boolean isDone = currentTaskAsString.substring(4, 5).toUpperCase().equals("X");
```
Perhaps the arguments to .substring should be assigned variable names or have a comment explaining why those numbers were used?

xavierpok commented on src/main/java/functions/Load.java:
```diff
@@ -0,0 +1,102 @@
+                    int fromTimingStartIndex = currentTaskAsString.indexOf("(from:") + 7;
+                    int fromTimingEndIndex = currentTaskAsString.indexOf("to:")-1;
+                    int toTimingStartIndex = fromTimingEndIndex + 5;
+                    int toTimingEndIndex = currentTaskAsString.length() - 1;
```
Similarly, it might be difficult for new readers to understand why these numbers were used, although it might also be obvious it's just a matter of expected input.

Still, it might be good to remain clear!

xavierpok commented on src/main/java/functions/Load.java:
```diff
@@ -0,0 +1,102 @@
+
+        while (s.hasNext()) {
+            String currentTaskAsString = s.nextLine();
+            String taskType = currentTaskAsString.substring(1,2).toUpperCase();
```
I like the clever observation that you can tell a lot of task types just from their first names, wish I had thought of that!

However, it might cause issues down the line if just strings are used and passed around. Maybe we could convert to an enum? Maybe instead of acting on the output of this string, we could have a codeblock to parse it into an enum type, similar to the one you've defined in Parser :

```java
private enum ChatFunction {
        LIST,
        MARK,
        UNMARK,
        TODO,
        DEADLINE,
        EVENT,
        DELETE,
        FIND
    }
```

And maybe a code block like so :

```
ChatFunction func;
switch (taskString) {
 
                case "T":
                    func = ChatFunction.TODO;
                    break;
```

It might help if the code needs to be refactored later on!

xavierpok commented on src/main/java/functions/Parser.java:
```diff
@@ -0,0 +1,223 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import tasks.*;
```
Minor issue, but it might be best to avoid star imports

xavierpok commented on src/main/java/functions/Save.java:
```diff
@@ -0,0 +1,64 @@
+public class Save {
+    TaskList taskList;
+    private String saveFilePath;
+    private static final String tempFilePath = "temp.txt";
```
Maybe this constant could be in all-caps w/ underscores to seperate words?
```suggestion
    private static final String TEMP_FILE_PATH= "temp.txt";
```

xavierpok commented:
Overall, LGTM, just a few issues that might be better off if looked at!

nicrandomlee committed:
Merge remote-tracking branch 'origin/add-gradle-support' into add-gradle-support

nicrandomlee committed:
add JUnit

nicrandomlee committed:
update gradle dependencies

nicrandomlee committed:
Update Commands OOP

nicrandomlee committed:
update commands as string implementation

nicrandomlee committed:
Implement GUI

nicrandomlee committed:
Add GUI

nicrandomlee committed:
update minor bug

nicrandomlee committed:
Add javadoc

nicrandomlee committed:
Merge branch 'branch-A-JavaDoc'

nicrandomlee committed:
Add help extension

nicrandomlee committed:
edit parser

nicrandomlee committed:
Merge branch 'branch-Level-10'

nicrandomlee committed:
Update javadoc

nicrandomlee committed:
update javadoc

nicrandomlee committed:
abstract load command

nicrandomlee committed:
add assertions

nicrandomlee committed:
improve code quality

nicrandomlee committed:
add help

nicrandomlee committed:
Fix bugs

nicrandomlee committed:
Help command: Assist user in using available functions

A help command is user-friendly for new users who are unfamiliar with the application or are unsure of the available functions.

Let's,
- Include a new command to provide the list of available functions and how to use them

nicrandomlee committed:
Added JUnit test methods for deadline, event and todo.

There are several test cases to check the functionality of task classes.
The test checks for the correct string representation after the given
input is entered.

Include javadoc for all methods.

Added javadoc for all methods for better documentation and readability.

Added Assertions

Added assertions to check for null or invalid variable representations
to debug.

nicrandomlee committed:
Updated javadocs

Updated javadocs for all common classes, methods and constructors.

nicrandomlee committed:
Updated checkstyle criterion

Minor edits on checkstyle. Renamed package from "Runner" to "runner".

nicrandomlee committed:
Help command: Assist user in using available functions

A help command is user-friendly for new users who are unfamiliar with
the application or are unsure of the available functions.

Let's,
- Include a new command to provide the list of available functions
and how to use them

nicrandomlee committed:
Updated javadocs

Updated javadocs for all common classes, methods and constructors for
greater clarity of documentation and reader understandability.

nicrandomlee committed:
Added JUnit test cases

Test cases have been added for Parser to test for robustness
and edge cases for user prompts. Minor changes are added for
code quality purposes.

nicrandomlee committed:
Improve GUI

Updating the user and cupid bot icon to add more specialty
to bot.

nicrandomlee committed:
Refactor package

nicrandomlee committed:
Update Readme

Added more details to readme as user guide.

nicrandomlee committed:
Update Readme

Provides user with more context of starting out, launching and using
this bot.

nicrandomlee committed:
Bug fixes

Added support for datetime edge case.

nicrandomlee committed:
Add resource

nicrandomlee committed:
Update Readme

Sample screenshot of UI included.

nicrandomlee committed:
Alter file path

Change file path of items

nicrandomlee committed:
Update Readme

Updated new link to Ui image.

nicrandomlee committed:
Update Readme

nicrandomlee committed:
Update Readme

nicrandomlee committed:
Update Readme

nicrandomlee committed:
Merge branch 'master' into branch-A-Assertions

nicrandomlee committed:
Update save

nicrandomlee committed:
Update Save

nicrandomlee committed:
Merge pull request #2 from nicrandomlee/branch-A-Assertions

Update save

nicrandomlee committed:
Update save

nicrandomlee committed:
Update Readme
